lib/cylc/task_events_mgr.py

```diff
@@ -287,77 +294,77 @@ def process_message(self, itask, priority, message, poll_event_time=None,
+        elif message == TASK_OUTPUT_FAILED:
+            if (itask.state.is_greater_than(TASK_STATUS_FAILED) and not
+                    itask.state.confirming_with_poll):
+                itask.state.confirming_with_poll = True
+                poll_func(self.suite, [itask], msg=poll_msg)
+                return
+            itask.state.confirming_with_poll = False
             self._process_message_failed(itask, event_time, self.JOB_FAILED)
         elif message == self.EVENT_SUBMIT_FAILED:
+            if (itask.state.is_greater_than(TASK_STATUS_SUBMIT_FAILED) and not
+                    itask.state.confirming_with_poll):
+                itask.state.confirming_with_poll = True
+                poll_func(self.suite, [itask], msg=poll_msg)
+                return
+            itask.state.confirming_with_poll = False
             self._process_message_submit_failed(itask, event_time)
         elif message == TASK_OUTPUT_SUBMITTED:
+            if (itask.state.is_greater_than(TASK_STATUS_SUBMITTED) and not
+                    itask.state.confirming_with_poll):
+                itask.state.confirming_with_poll = True
+                poll_func(self.suite, [itask], msg=poll_msg)
+                return
+            itask.state.confirming_with_poll = False
             self._process_message_submitted(itask, event_time)
         elif message.startswith(TaskMessage.FAIL_MESSAGE_PREFIX):
-            # Task failed message
+            # Task received signal.
             signal = message[len(TaskMessage.FAIL_MESSAGE_PREFIX):]
             self._db_events_insert(itask, "signaled", signal)
             self.suite_db_mgr.put_update_task_jobs(
                 itask, {"run_signal": signal})
+            if (itask.state.is_greater_than(TASK_STATUS_FAILED) and not
+                    itask.state.confirming_with_poll):
+                itask.state.confirming_with_poll = True
+                poll_func(self.suite, [itask], msg=poll_msg)
+                return
+            itask.state.confirming_with_poll = False
             self._process_message_failed(itask, event_time, self.JOB_FAILED)
         elif message.startswith(TaskMessage.ABORT_MESSAGE_PREFIX):
             # Task aborted with message
-            signal = message[len(TaskMessage.ABORT_MESSAGE_PREFIX):]
+            aborted_with = message[len(TaskMessage.ABORT_MESSAGE_PREFIX):]
             self._db_events_insert(itask, "aborted", message)
             self.suite_db_mgr.put_update_task_jobs(
-                itask, {"run_signal": signal})
-            self._process_message_failed(itask, event_time, signal)
+                itask, {"run_signal": aborted_with})
+            if (itask.state.is_greater_than(TASK_STATUS_FAILED) and not
+                    itask.state.confirming_with_poll):
+                itask.state.confirming_with_poll = True
+                poll_func(self.suite, [itask], msg=poll_msg)
+                return
```

Comment:
We have a block here that is repeated multiple times with only 1 difference. I wonder if it is worth moving this little bit of logic into a separate private method or not.

Reply:
Yes, agreed (I recall considering that at the time, but forgot to come back to it...)